web/source/kmwcallback.ts

```diff
@@ -7,13 +7,65 @@
    Copyright 2017 SIL International
 ***/
 
+/*
+ * Type alias definitions to reflect the parameters of the fullContextMatch() callback (KMW 10+).
+ * No constructors or methods since keyboards will not utilize the same backing prototype, and
+ * property names are shorthanded to promote minification.
+ */
+
+type PlainKeyboardStore = string;
+
+// TODO:  Implement the new 'store object-orientation proposal.
+//type ComplexKeyboardStore = (string|{d: number})[]; 
+
+type KeyboardStore = PlainKeyboardStore; // | ComplexKeyboardStore;
+type RuleChar = string;
+
+class RuleDeadkey {
+  /**
+   * Existence of this property indicates the object is a deadkey.  
+   * 
+   * Value:  the deadkey's ID.
+   */
+  ['d']: number; // For 'd'eadkey; also reflects the Deadkey class's 'd' property.
+}
+
+class ContextAny {
+  /**
+   * Existence of this property indicates the object reflects an "any" statement.  
+   * 
+   * Value:  the store to search.
+   */
+  ['a']: KeyboardStore; // For 'a'ny statement.
+}
+
+class RuleIndex {
+  /**
+   * Existence of this property indicates the object reflects an "index" statement.  
+   * 
+   * Value:  {'s': the Store from which to output, 'o': the Offset into the current rule's context.}
+   */
+  ['i']: {'s': KeyboardStore, 'o': number}; // For 'i'ndex statement.
+}
+
+class ContextEx {
+  /**
+   * Existence of this property indicates that the object reflects a "context" Keyman language statement.
+   * 
+   * Value:  The offset into the current rule's context to be matched.
+   */
+  ['c']: number; // For 'c'ontext statement.
+}
+
+type ContextEntry = RuleChar | RuleDeadkey | ContextAny | RuleIndex | ContextEx;
```

Comment:
I think I'd like to see this constructed as a discriminated union -- it makes type guards simpler later on and allows us to build in compile-time checks for completeness where we work with it; see https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions

Reply:
Note that using a discriminated union would require Developer to compile in an extra parameter for each object that appears in a `KFCM` rule, as we'd need to add in a _discriminant_ property.  This will, of course, increase the size of v10+ keyboards.  Are we okay with this?

Reply:
Good point. That may be a good enough reason to not do it.